src/test/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStoreTest.java

```diff
@@ -54,73 +63,314 @@
             .childHandler(
-                new ChannelInitializer<NioSocketChannel>() {
+                new ChannelInitializer<Channel>() {
                   @Override
-                  protected void initChannel(NioSocketChannel ch) {
+                  protected void initChannel(Channel ch) {
                     ch.pipeline().addLast(new HttpServerCodec());
                     ch.pipeline().addLast(new HttpObjectAggregator(1000));
                     ch.pipeline().addLast(handler);
                   }
                 });
-    return ((ServerSocketChannel) sb.bind("localhost", 0).sync().channel());
+    try {
+      return ((ServerChannel) sb.bind(socketAddress).sync().channel());
+    } catch (Exception e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  private static DomainSocketAddress newDomainSocketAddress() {
+    try {
+      File file = File.createTempFile("bazel", ".sock");
+      file.delete();
+      return new DomainSocketAddress(file.getAbsoluteFile());
+    } catch (Exception e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  interface TestServer {
+
+    ServerChannel start(ChannelInboundHandler handler);
+
+    void stop(ServerChannel serverChannel);
+  }
+
+  private static final class InetTestServer implements TestServer {
+
+    public ServerChannel start(ChannelInboundHandler handler) {
+      return createServer(
+          NioServerSocketChannel.class,
+          NioEventLoopGroup::new,
+          new InetSocketAddress("localhost", 0),
+          handler);
+    }
+
+    public void stop(ServerChannel serverChannel) {
+      try {
+        serverChannel.close();
+        serverChannel.closeFuture().sync();
+        serverChannel.eventLoop().shutdownGracefully().sync();
+      } catch (Exception e) {
+        throw new IllegalStateException(e);
+      }
+    }
+  }
+
+  private static final class UnixDomainServer implements TestServer {
+
+    // Note: this odd implementation is a workaround because we're unable to shut down and restart
+    // KQueue backed implementations. See https://github.com/netty/netty/issues/7047.
+
+    private final ServerChannel serverChannel;
+    private ChannelInboundHandler handler = null;
+
+    public UnixDomainServer(
+        Class<? extends ServerChannel> serverChannelClass,
+        IntFunction<EventLoopGroup> newEventLoopGroup
+    ) {
+      EventLoopGroup eventLoop = newEventLoopGroup.apply(1);
+      ServerBootstrap sb =
+          new ServerBootstrap()
+              .group(eventLoop)
+              .channel(serverChannelClass)
+              .childHandler(
+                  new ChannelInitializer<Channel>() {
+                    @Override
+                    protected void initChannel(Channel ch) {
+                      ch.pipeline().addLast(new HttpServerCodec());
+                      ch.pipeline().addLast(new HttpObjectAggregator(1000));
+                      ch.pipeline().addLast(new ChannelInboundHandler() {
```

Comment:
I think this can be simplified to just `.addLast(Preconditions.checkNotNull(handler))`. The `initChannel` method won't be called until after `UnixDomainServer.start(handler)` no?